prog/minio/minio_cluster_nexus.rb

```diff
@@ -0,0 +1,95 @@
+# frozen_string_literal: true
+
+require_relative "../../lib/util"
+
+class Prog::Minio::MinioClusterNexus < Prog::Base
+  subject_is :minio_cluster
+
+  semaphore :restart, :destroy
+
+  def self.assemble(project_id, cluster_name, location, admin_user,
+    storage_size_gib, pool_count, server_count, driver_count, vm_size)
+    unless (project = Project[project_id])
+      fail "No existing project"
+    end
+
+    Validation.validate_vm_size(vm_size)
+    Validation.validate_location(location, project.provider)
+    Validation.validate_name(cluster_name)
+    Validation.validate_minio_username(admin_user)
+
+    DB.transaction do
+      ubid = MinioCluster.generate_ubid
+
+      minio_cluster = MinioCluster.create(
+        name: cluster_name,
+        location: location,
+        admin_user: admin_user,
+        admin_password: SecureRandom.base64(15).gsub(/[+\/]/, "+" => "_", "/" => "-"),
+        target_total_storage_size_gib: storage_size_gib,
+        target_total_pool_count: pool_count,
+        target_total_server_count: server_count,
+        target_total_driver_count: driver_count,
+        target_vm_size: vm_size
+      ) { _1.id = ubid.to_uuid }
+      minio_cluster.associate_with_project(project)
+
+      Strand.create(prog: "Minio::MinioClusterNexus", label: "start") { _1.id = minio_cluster.id }
+    end
+  end
+
+  def before_run
+    when_destroy_set? do
+      if strand.label != "destroy"
+        hop_destroy
+      end
+    end
+  end
+
+  label def start
+    register_deadline(:wait, 10 * 60)
+    subnet_st = Prog::Vnet::SubnetNexus.assemble(
+      Config.minio_service_project_id,
+      name: "#{minio_cluster.name}-subnet",
+      location: minio_cluster.location
+    )
+    minio_cluster.update(private_subnet_id: subnet_st.id)
+
+    minio_cluster.target_total_pool_count.times do |i|
+      Prog::Minio::MinioPoolNexus.assemble(minio_cluster.id, i * minio_cluster.per_pool_server_count)
+    end
+
+    hop_wait_pools
+  end
+
+  label def wait_pools
+    if minio_cluster.minio_pools.all? { _1.waiting? }
```

Comment:
Feel like we gotta do something about these polling loops. Maybe not this patch, though. Basically supervising strands need a way to install a kind of hook and then go to sleep, and let the supervised strand wake em up. You could think of it as a condition variable.